pkg/agent/plugin/svidstore/gcpsecretmanager/gcloud.go

```diff
@@ -0,0 +1,286 @@
+package gcpsecretmanager
+
+import (
+	"context"
+	"crypto/sha1" //nolint: gosec // We use sha1 to hash TD in 128b to avoid label restrictions
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/hcl"
+	configv1 "github.com/spiffe/spire-plugin-sdk/proto/spire/service/common/config/v1"
+	"github.com/spiffe/spire/pkg/agent/plugin/svidstore"
+	"github.com/spiffe/spire/pkg/common/catalog"
+	svidstorev1 "github.com/spiffe/spire/proto/spire/plugin/agent/svidstore/v1"
+	secretmanagerpb "google.golang.org/genproto/googleapis/cloud/secretmanager/v1"
+	"google.golang.org/genproto/googleapis/iam/v1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	pluginName = "gcp_secretmanager"
+)
+
+func BuiltIn() catalog.BuiltIn {
+	return builtin(New())
+}
+
+func builtin(p *SecretManagerPlugin) catalog.BuiltIn {
+	return catalog.MakeBuiltIn(pluginName,
+		svidstorev1.SVIDStorePluginServer(p),
+		configv1.ConfigServiceServer(p),
+	)
+}
+
+func New() *SecretManagerPlugin {
+	return newPlugin(newSecretManagerClient)
+}
+
+func newPlugin(newClient func(context.Context, string) (secretManagerClient, error)) *SecretManagerPlugin {
+	p := &SecretManagerPlugin{}
+	p.hooks.newClient = newClient
+
+	return p
+}
+
+type Configuration struct {
+	ServiceAccountFile string   `hcl:"service_account_file" json:"service_account_file"`
+	UnusedKeys         []string `hcl:",unusedKeys" json:",omitempty"`
+}
+
+type SecretManagerPlugin struct {
+	svidstorev1.UnsafeSVIDStoreServer
+	configv1.UnsafeConfigServer
+
+	log    hclog.Logger
+	mtx    sync.RWMutex
+	client secretManagerClient
+	tdHash string
+
+	hooks struct {
+		newClient func(context.Context, string) (secretManagerClient, error)
+	}
+}
+
+func (p *SecretManagerPlugin) SetLogger(log hclog.Logger) {
+	p.log = log
+}
+
+// Configure configures the SecretsMangerPlugin.
+func (p *SecretManagerPlugin) Configure(ctx context.Context, req *configv1.ConfigureRequest) (*configv1.ConfigureResponse, error) {
+	// Parse HCL config payload into config struct
+	config := &Configuration{}
+	if err := hcl.Decode(config, req.HclConfiguration); err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "unable to decode configuration: %v", err)
+	}
+
+	if len(config.UnusedKeys) != 0 {
+		return nil, status.Errorf(codes.InvalidArgument, "unknown configurations detected: %s", strings.Join(config.UnusedKeys, ","))
+	}
+
+	client, err := p.hooks.newClient(ctx, config.ServiceAccountFile)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to create secretmanager client: %v", err)
+	}
+
+	// gcp secret manager does not allow ".", hash td as label
+	tdHash := sha1.Sum([]byte(req.CoreConfiguration.TrustDomain)) //nolint: gosec // We use sha1 to hash TD in 128b to avoid label restrictions
```

Comment:
The limitations of the label strings are really unfortunate.
Having a hash is not ideal in terms of user experience, since it's not understandable by humans looking at the secrets. I've been thinking about creative ways to work around this, but the limits in the characters set + the length limitation (63 characters) are really hard constraints to make this something readable by humans (trust domain names may have a length of 255).

@evan2645 any thoughts on this?

Reply:
> Having a hash is not ideal in terms of user experience, since it's not understandable by humans looking at the secrets.

I worried about the same

>  I've been thinking about creative ways to work around this, but the limits in the characters set + the length limitation (63 characters) are really hard constraints to make this something readable by humans (trust domain names may have a length of 255).

same :)

I don't have a better solution. If we do character substitution, we'd have to substitute with a character that's not otherwise allowed in the trust domain name. And to address the length problem, we could truncate ... I'm not sure how real the risk of collision is here ... but then again, it's certainly less risky than not scoping the label to trust domain name.